src/lsm/forest.zig

```diff
@@ -155,101 +203,205 @@ pub fn ForestType(comptime Storage: type, comptime groove_cfg: anytype) type {
                 @field(forest.grooves, field.name).reset();
             }
 
+            forest.manifest_log.reset();
             forest.node_pool.reset();
 
             forest.* = .{
                 // Don't reset the grid – replica is responsible for grid cancellation.
                 .grid = forest.grid,
                 .grooves = forest.grooves,
                 .node_pool = forest.node_pool,
+                .manifest_log = forest.manifest_log,
             };
         }
 
-        fn JoinType(comptime join_op: JoinOp) type {
-            return struct {
-                pub fn start(forest: *Forest, callback: Callback) void {
-                    assert(forest.join_op == null);
-                    assert(forest.join_pending == 0);
-                    assert(forest.join_callback == null);
-
-                    forest.join_op = join_op;
-                    forest.join_pending = std.meta.fields(Grooves).len;
-                    forest.join_callback = callback;
+        pub fn open(forest: *Forest, callback: Callback) void {
+            assert(forest.progress == null);
+            forest.progress = .{ .open = .{ .callback = callback } };
+
+            inline for (std.meta.fields(Grooves)) |field| {
+                @field(forest.grooves, field.name).open_commence(&forest.manifest_log);
+            }
+
+            forest.manifest_log.open(manifest_log_open_event, manifest_log_open_callback);
+        }
+
+        fn manifest_log_open_event(
+            manifest_log: *ManifestLog,
+            level: u7,
+            table: *const schema.ManifestLog.TableInfo,
+        ) void {
+            const forest = @fieldParentPtr(Forest, "manifest_log", manifest_log);
+            assert(forest.progress.? == .open);
+            assert(level < constants.lsm_levels);
+
+            // TODO Is there a faster (or cleaner) way to map tree-id to the tree?
```

Comment:
Wasn't able to do a working poc in a few minutes, but I think something of this shape should be possible: 

```
/// Given a comptime tree id, lookup the tree type.
/// This is basically lookup in tree_infos
fn TreeType(tree_id: comptime_int) type

/// Given a comptime tree id and a runtime forset, return a pointer to the tree
/// This is `inline for` we already we have in manifest_log_open
fn tree(forest: *Forest, tree_id: comptime_int) *TreeType(tree_id) {}

/// The largest tree ID we have, for loop over tree infos.
const tree_id_max: comptime_int = ...

fn manifest_log_open_event(table) {
    switch (table.tree_id) {
        // key idea -- covert to comptime-known tree id here
        inline 1..tree_id_max => |tree_id| tree(forest, tree_id).open_table()
        else => @panic()
    }

}

```

Reply:
Does that make a difference though? I would expect `switch (...) { inline 1..tree_id_max => ... }` to compile down to essentially the same chain of `if (id == 1) { ... }; if (id == 2) { ... } etc`.

Reply:
I think inline case is guaranteed to be a jump table. Here's a debug example in godbolt, where `inline for` generates an if-chain, and `inline case` gets a jump table (in release for this small example, both get a jumptable):

https://godbolt.org/z/Eq5EcsPPn